src/control/character_controller.rs

```diff
@@ -251,25 +246,32 @@ impl KinematicCharacterController {
                     toi,
                 });
 
-                if let (Some(translation_on_slope), _) =
-                    self.handle_slopes(&toi, &mut translation_remaining)
-                {
-                    translation_remaining = translation_on_slope;
-                } else {
-                    // If the slope is too big, try to step on the stair.
-                    self.handle_stairs(
-                        bodies,
-                        colliders,
-                        queries,
-                        character_shape,
-                        &(Translation::from(result.translation) * character_pos),
-                        &dims,
-                        filter,
-                        handle,
-                        &mut translation_remaining,
-                        &mut result,
-                    );
+                // Try to go up stairs.
+                if !self.handle_stairs(
+                    bodies,
+                    colliders,
+                    queries,
+                    character_shape,
+                    &(Translation::from(result.translation) * character_pos),
+                    &dims,
+                    filter,
+                    handle,
+                    &mut translation_remaining,
+                    &mut result,
+                ) {
+                    // No stairs, try to move along slopes.
+                    if let Some(translation_on_slope) =
+                        self.handle_slopes(&toi, &mut translation_remaining, offset)
+                    {
+                        translation_remaining = translation_on_slope;
+                    } else {
+                        // No slopes or stairs ahead; try to move along obstacle.
+                        let allowed_translation = subtract_hit(translation_remaining, &toi, offset);
+                        result.translation += allowed_translation;
+                        translation_remaining -= allowed_translation;
```

Comment:
I may be missing something but this looks too optimistic. Not having any slope or step ahead does not imply that we can move along the obstacle by the complete `allowed_translation` since there could be other obstacles ahead.
In addition, doesn’t that imply that this `allowed_translation` would be applied twice? (It was already applied at line 238).

Reply:
Hmm, I see. For the fix, I noticed that often it's not a second time, but a first time. The line you mention stops the movement completely in some cases, e.g. when trying to jump up while moving up the wall on the right. How about we run this code only when it would be the first we move along the obstacle? See [`760ea8d` (#446)](https://github.com/dimforge/rapier/pull/446/commits/760ea8d498c7dfc99d59342c8ec492de193ffed9)